### DISCUSSION TITLE
Build Error: Could not find the sentinel NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2

### DISCUSSION THREAD

**Discussion by Tweekism** [jannis-baum/Vivify#80] [Q&A] [answered]
Hey Jannis,

Thought best I should start by trying to build it. Are you able to offer any advice on my build error here?

Google suggest it might be something to do with versions of node > v20.0.0

```sh
 $ node --version
v20.12.2
```

```sh
 $ yarn
yarn install v1.22.22
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...

Done in 9.75s.
  $ ./configure ~/.local/bin
  $ make install
mkdir -p build
rm -rf build/static.zip
zip -X -r build/static.zip static
  adding: static/ (stored 0%)
  adding: static/client.js (deflated 55%)
  # ...
  # Snip: ...A whole bunch of statics...
  # ...
  adding: static/katex/katex.css (deflated 85%)
  adding: static/style.css (deflated 75%)
node_modules/.bin/webpack
asset bundle.js 3.26 MiB [emitted] [minimized] (name: main) 1 related asset
orphan modules 9.85 KiB [orphan] 15 modules
runtime modules 1.04 KiB 5 modules
modules by path ./node_modules/ 4.64 MiB
  javascript modules 4.32 MiB 410 modules
  json modules 324 KiB 13 modules
modules by path ./src/ 17.4 KiB
  modules by path ./src/parser/*.ts 4.51 KiB 4 modules
  modules by path ./src/routes/*.ts 7.66 KiB 3 modules
  modules by path ./src/*.ts 3.94 KiB 2 modules
  + 1 module
optional modules 126 bytes [optional]
  external "async_hooks" 42 bytes [optional] [built] [code generated]
  external "bufferutil" 42 bytes [optional] [built] [code generated]
  external "utf-8-validate" 42 bytes [optional] [built] [code generated]
+ 21 modules

WARNING in ./node_modules/express/lib/view.js 81:13-25
Critical dependency: the request of a dependency is an expression
 @ ./node_modules/express/lib/application.js 22:11-28
 @ ./node_modules/express/lib/express.js 18:12-36
 @ ./node_modules/express/index.js 11:0-41
 @ ./src/app.ts 17:34-52

1 warning has detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

webpack 5.92.1 compiled with 1 warning in 8570 ms
touch build/bundle.js
mkdir -p build/linux
rm -rf build/linux/vivify-server
node --experimental-sea-config sea-config.json
Wrote single executable preparation blob to build/sea-prep.blob
cp /usr/bin/node build/linux/vivify-server
chmod +w build/linux/vivify-server
node_modules/.bin/postject build/linux/vivify-server NODE_SEA_BLOB build/sea-prep.blob \
		  --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2
Start injection of NODE_SEA_BLOB in build/linux/vivify-server...
warning: Can't find string offset for section name '.note.cafe1a7e'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note'
Error: Could not find the sentinel NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2 in the binary
make: *** [Makefile:69: build/linux/vivify-server] Error 1
```

**Comment by jannis-baum** [comment]
Hello!

We recently switched to [Node SEA](https://nodejs.org/api/single-executable-applications.html) to compile Vivify into a single executable. This is still in beta, which is why there seem to be some build issues on some systems. We haven't seen this exact error yet. Could you verify your version again by adding the following to the Makefile (or a new one) and then running `make node-version`?

```Makefile
.PHONY: node-version
node-version:
	$(shell command -v node) --version
```

My first idea based on the error message is that the version of node that we try to inject into is different, maybe older, than what you get with `node --version`.

If it's not that, I would suggest that you either use the CI builds until there are updates to Node SEA, or, if you really want to build it yourself, go through the example on the [docs](https://nodejs.org/api/single-executable-applications.html) and see if you can get that to work, then we can figure out what the problem here could be :)

Also note that for developing Vivify you don't actually need to be able to build it yourself: You can use `yarn dev` to try it out with live-updates as you are coding, and then as soon as you push something to the repo the CI will build the current version for you, so you can install it from there

**Reply by Tweekism** [reply]
> Also note that for developing Vivify you don't actually need to be able to build it yourself: You can use `yarn dev` to try it out with live-updates as you are coding, and then as soon as you push something to the repo the CI will build the current version for you, so you can install it from there

Good to know, I'll def look into that as well. For Fedora packaging I will need to be able to build it, as you can't submit binaries there, only source. And I'm more than happy to help work out and build issues.

**Reply by jannis-baum** [reply]
Okay, hope we can figure this out then. For @tuurep we have also had problems building/executing the build (reference https://github.com/jannis-baum/vivify/pull/68#issuecomment-2214118038 and following) that we still haven't been able to solve. I think it'll be good to see how it goes with the docs example for you and if that doesn't work, we should open issues [here](https://github.com/nodejs/single-executable/issues?q=is%3Aopen+is%3Aissue).

**Reply by Tweekism** [reply]
```sh
 $ make node-version
/usr/bin/node --version
v20.12.2
```
![image](https://github.com/user-attachments/assets/8eb25d32-81fc-47b8-854b-4ff37c2cf8e1)

**Reply by Tweekism** [reply]
Something to note. Reverting to node v20.0.0 does then build successfully (with a few warning, see below).

Did you want me to create an issue for this?

```sh 
 $ make clean
rm -rf build/*
 $ make node-version
/home/jason/.nvm/versions/node/v20.0.0/bin/node --version
v20.0.0
 $ make install
mkdir -p build
rm -rf build/static.zip
zip -X -r build/static.zip static
  adding: static/ (stored 0%)
# ...
# Snip
# ...
  external "utf-8-validate" 42 bytes [optional] [built] [code generated]
+ 21 modules

WARNING in ./node_modules/express/lib/view.js 81:13-25
Critical dependency: the request of a dependency is an expression
 @ ./node_modules/express/lib/application.js 22:11-28
 @ ./node_modules/express/lib/express.js 18:12-36
 @ ./node_modules/express/index.js 11:0-41
 @ ./src/app.ts 17:34-52

1 warning has detailed information that is not shown.
Use 'stats.errorDetails: true' resp. '--stats-error-details' to show it.

webpack 5.92.1 compiled with 1 warning in 7697 ms
touch build/bundle.js
mkdir -p build/linux
rm -rf build/linux/vivify-server
node --experimental-sea-config sea-config.json
Wrote single executable preparation blob to build/sea-prep.blob
cp /home/jason/.nvm/versions/node/v20.0.0/bin/node build/linux/vivify-server
chmod +w build/linux/vivify-server
node_modules/.bin/postject build/linux/vivify-server NODE_SEA_BLOB build/sea-prep.blob \
		  --sentinel-fuse NODE_SEA_FUSE_fce680ab2cc467b6e072b8b5df1996b2
Start injection of NODE_SEA_BLOB in build/linux/vivify-server...
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note.100'
warning: Can't find string offset for section name '.note.100'
💉 Injection done!
chmod -w build/linux/vivify-server
rm -rf /home/jason/.local/bin/vivify-server
cp build/linux/vivify-server /home/jason/.local/bin/vivify-server
mkdir -p build/linux
cp viv build/linux/viv
rm -rf /home/jason/.local/bin/viv
cp build/linux/viv /home/jason/.local/bin/viv

```

**Reply by Tweekism** [reply]
Ok, turns out there is more going on, I have to head out again for a bit, but basically if you use the Fedora packaged version of node it fails, but if you use the same version number but pull it from npm it works fine.


```sh
 $ nvm install 20.12.2
Downloading and installing node v20.12.2...
Downloading https://nodejs.org/dist/v20.12.2/node-v20.12.2-linux-x64.tar.xz...
################################################################ 100.0%
Computing checksum with sha256sum
Checksums matched!
Now using node v20.12.2 (npm v10.5.0)
 $ make clean
rm -rf build/*
 $ make install
```

**^^ Works fine**

```sh
 $ nvm use system
Now using system version of node: v20.12.2 (npm v10.5.0)
 $ make clean
rm -rf build/*
 $ make install
```

**^^ Fails**

So could just be something Fedora is doing to their version that they package.

I'll look into it a bit more later and then raise an issue with the all deets.

**Reply by jannis-baum** [reply]
That's interesting! Good observations. Might come down to what libraries the node executables link to as well like with @tuurep. What do you get if you run

```sh
objdump -p path/to/the/node/executable | grep NEEDED
```

for the executable it works with vs. for the one it doesn't?

---

@tuurep maybe using a `node` version from `nvm` will make the build from your system work as well?

**Reply by Tweekism** [reply]
I'll go read that thread you mention with **tuurep**, in the meantime:-

From Fedora repo

```sh
 $ which node
/usr/bin/node
 $ objdump -p /usr/bin/node | grep NEEDED
  NEEDED               libnode.so.115
  NEEDED               libc.so.6
```

From https://nodejs.org/dist/v20.12.2/

```sh
 $ which node
~/.nvm/versions/node/v20.12.2/bin/node
 $ objdump -p ~/.nvm/versions/node/v20.12.2/bin/node | grep NEEDED
  NEEDED               libdl.so.2
  NEEDED               libstdc++.so.6
  NEEDED               libm.so.6
  NEEDED               libgcc_s.so.1
  NEEDED               libpthread.so.0
  NEEDED               libc.so.6
  NEEDED               ld-linux-x86-64.so.2
```

**Reply by jannis-baum** [reply]
Ah, yes, starting to see a pattern here. Seems like if the output is different from the one you got from `~/.nvm/versions/node/v20.12.2/bin/node` it will fail.

The closest thing to documentation I found on this was from https://github.com/nodejs/single-executable/issues/46#issue-1395392448:

> Node.js links to GLIBC. Depending on the official release configuration, building against too new GLIBC versions reduces Linux compatibility

So maybe it will come down to whoever makes the packages for Fedora, Arch, and anywhere else it fails, realizing this issue and then having the motivation to take care of it?

---

Anyways, I guess we have preliminarily resolved this now that you found that it works with the `nvm` releases? At least in terms of Vivify's area of concern. I'll be interested to see if this also fixes it for @tuurep. I would suggest that once he confirms we can add a note on it to the `CONTRIBUTING.md` and then close this

**Comment by Tweekism** [comment]
Sounds good to me, I wonder what it'll mean when it comes time to package for Fedora, but I'll let you know when I get there.

@tuurep you familiar with node version manager?

**Comment by tuurep** [comment]
Hi, noted and getting back to testing this but can't right now :)

**Reply by Tweekism** [reply]
I'm heading to bed now, but if you need any clarification on the steps I did, just give me a yell :)

**Comment by tuurep** [comment]
Hey @jannis-baum @Tweekism yes this works!

I used

```
nvm install node
```

where `node` means latest version (22.4.1) so that's the same as pacman repos.

Thanks Tweekism this solves a pretty big problem we've had here! Finally I'm able to build for myself too :smile:

**Reply by Tweekism** [reply]
No worries mate, I'm just doing doing this again now so I can build on my work PC.

I'll create an issue in the tracker to update our contribution help file to include this.

**Comment by jannis-baum** [comment, marked as answer]
Nice, good work on figuring this out! I will close this discussion now then. Just to summarize:

The error happens because some Linux distributions package `node` with links to different shared libraries than what is required for Node SEA to work. To fix it, you can, for example, use a node version from [`nvm`](https://github.com/nvm-sh/nvm). @Tweekism is making sure this is documented in #91.
